### PR TITLE
disambiguate assembly chains by appending assembly id

### PR DESCRIPTION
**Pull request by alex-hh**
Biotite doesn't currently distinguish between the chains produced by symmetry operations on a given asym unit in an assembly.

This can make certain selection operations slightly inconvenient (e.g. getting residue starts of a specific chain etc.) and is inconsistent with the chain naming convention employed by the PDB in assembly .cif files.

This PR follows the PDB convention of appending a '-<oper_id>' to the chain ids.

### CODE REVIEW

**Comment by padix-key** [comment]
Hi. First I thought that this is the same as #389, but while separating chains is also possible without changing the `AtomArray` annotations, this is not true if one wants to have all chains for a specific symmetry operation. So I see definitely the use case for this feature.

However, I am leaning towards keeping the original `chain_id` annotation here, as it represents the `asym_id` from the `atom_site` category. I think in the assembly CIF files they go this route, because they have no other way of representing this while keeping the file compatible with the PDBx dictionary.

Instead I propose to add an additional optional annotation array (what do you think about `sym_id`?) that tracks this information instead. Similar to the appended number in your draft it would go from 0 to N. If a user still wants to have this information in the `chain_id`, they could concatenate `chain_id` with `sym_id` via [`numpy.strings.add()`](https://numpy.org/doc/stable/reference/generated/numpy.strings.add.html).

As a pointer, have a look at `set_annotation()` in the [`AtomArray`](https://www.biotite-python.org/latest/apidoc/biotite.structure.AtomArray.html) class, for how to add a custom annotation.

**Comment by Croydon-Brixton** [comment]
I'd also be in favour of a separate `sym_id` annotation.

The one caveat I'd have is that some symmetry operations are actually 'slicing' type operations rather than copying operations, so if we wanted to add a `sym_id` to the asym unit we would need to decide whether we assign sym_id's based on the operation id or whether we want to assign them based on actual chemical similarity. (See `5ocm` for an example). Personally I'd lean towards just using the oper_id as you suggested, @alex-hh

**Comment by alex-hh** [comment]
Hi - Thanks for the pointer to the previous issue, this also helps with some of the use cases I was considering.

Making sym_id an optional annotation makes sense to me.

I've updated the PR with one possible solution for the sym_id (concatenation of the oper ids applied to generate the sym unit), but agree it's not obvious what the value should be. I was trying to follow the convention from the assembly.cif files, but I'm not sure what they do in complex cases there.

**Comment by padix-key** [comment]
Having `sym_id` concatenate the different operation IDs would also be OK for me. I suspect there are no strctures where a high number of these operations are combined

File: src/biotite/structure/io/pdbx/convert.py [hunk 1]
```diff
@@ -1481,6 +1481,7 @@ def get_assembly(
     extra_fields=None,
     use_author_fields=True,
     include_bonds=False,
+    include_sym_id=False,
```
Comment (padix-key) [inline src/biotite/structure/io/pdbx/convert.py:1484]: I think we do not need this additional parameter: The extra annotation has only a small overhead, so I think it is OK to always include the `sym_id`.

File: src/biotite/structure/io/pdbx/convert.py [hunk 2]
```diff
@@ -1658,9 +1665,16 @@ def _apply_transformations(structure, transformation_dict, operations):
             coord = matrix_rotate(coord, rotation_matrix)
             # Translate
             coord += translation_vector
+
+        assembly_transform_ids.append(
+            np.full(len(structure), "-".join(list(operation)))
+        )
```
Comment (padix-key) [inline src/biotite/structure/io/pdbx/convert.py:1671]: For performance reasons, I think it would be better to append single values here and repeat the array below.

```suggestion
        sym_ids.append("-".join(list(operation)))
```

File: src/biotite/structure/io/pdbx/convert.py [hunk 3]
```diff
@@ -1658,9 +1665,16 @@ def _apply_transformations(structure, transformation_dict, operations):
...
         assembly_coord[i] = coord
 
-    return repeat(structure, assembly_coord)
+    assembly = repeat(structure, assembly_coord)
+    if include_sym_id:
+        assembly.set_annotation("sym_id", np.concatenate(assembly_transform_ids))
```
Comment (padix-key) [inline src/biotite/structure/io/pdbx/convert.py:1676]: See above

```suggestion
        assembly.set_annotation(
            "sym_id", np.repeat(assembly_transform_ids, structure.array_length)
        )
```

**Review by padix-key — CHANGES_REQUESTED** [review]
Hi, I would recommend a few small changes, but overall it looks good!

**Comment by alex-hh** [comment]
updated!

**Comment by padix-key** [comment]
I took the freedom to add some tests and documentation about the new annotation. Have a look of you like. From my side this would be ready to merge, when the discussion above is resolved.

File: src/biotite/structure/io/pdbx/convert.py [hunk 4]
```diff
@@ -1658,9 +1658,13 @@ def _apply_transformations(structure, transformation_dict, operations):
             coord = matrix_rotate(coord, rotation_matrix)
             # Translate
             coord += translation_vector
+
+        sym_ids.append("-".join(list(operation)))
         assembly_coord[i] = coord
 
-    return repeat(structure, assembly_coord)
+    assembly = repeat(structure, assembly_coord)
+    assembly.set_annotation("sym_id", np.repeat(sym_ids, structure.array_length()))
```
[the remark this answers is not in the document]
Comment (Croydon-Brixton) [reply src/biotite/structure/io/pdbx/convert.py:1669]: @padix-key @alex-hh 
Sorry for taking longer to get back -- and thank you for this, this is very helpful!

**TL;DR**: I still believe an integer encoding would be a better design decision here that'd be just as easy to implement, but agree that this is a nit-picky point that may only be relevant in a handful of cases. Overall I'd be happy with the current implementation.

- `4v5x`: you're right - bad choice by me this one indeed only has one assembly (but a ton of chains :D) 
- `6o3h`: Thank you for going in depth with this example. I agree that `"60"` would not be super concerning. 
- `1f2n`: This was the type of behaviour I was worried about, thank you for identifying one of these cases. For this particular one, the string only gets up to length `"X0-60"` , which isn't terrible but still pretty hefty compared to what it could be: stored e.g. via numpy's `<U` encoding -- utf32, i.e. 4 bytes iirc -- this would be `5*4=20` bytes per element. That's more than the `3*4=12` (float 32) bytes per coordinate in this case. Instead, with an integer encoding this could easily be `1` byte (uint8) or `2` bytes (uint16). This would make the atom-array both, more lightweight and comparisons over the `sym_id` dimension faster because we can do int comparisons rather than string comparisons.

What I don't know is if there are cases where this id could become even longer? (@padix-key or @nscorley, maybe you know more?). In the past I've often been surprised by obscure edge cases in the PDB that I thought would never happen, so nowadays I usually try to rely on as little assumptions on PDB files as possible.

[the remark this answers is not in the document]
Comment (alex-hh) [reply src/biotite/structure/io/pdbx/convert.py:1669]: No I have no preference - I was just aiming for consistency with the PDB. (I also wanted to check that the examples you gave were in fact consistent).

I guess one consideration would be that in most cases the string will be an integer and it makes sense for this integer to be the same as the one used to identify the transformation in the cif file.

In cases where we compose transformations (it would be cool to have a proper complex example but I don't) I have no preference and don't even know what the PDB convention is.

[the remark this answers is not in the document]
Comment (padix-key) [reply src/biotite/structure/io/pdbx/convert.py:1669]: Okay, then I would settle with the integer counting up from `0` to `N`. I would also consistently always do it this way, even if the transformation is not composed of multiple operations, so the user knows what to expect.

@alex-hh Would you like to add the necessary changes, or should I do it?

[the remark this answers is not in the document]
Comment (alex-hh) [reply src/biotite/structure/io/pdbx/convert.py:1669]: Sounds good: if you're happy to do it it might be quicker! But also happy to if you don't have time

[the remark this answers is not in the document]
Comment (nscorley) [reply src/biotite/structure/io/pdbx/convert.py:1669]: > @padix-key @alex-hh Sorry for taking longer to get back -- and thank you for this, this is very helpful!
> 
> **TL;DR**: I still believe an integer encoding would be a better design decision here that'd be just as easy to implement, but agree that this is a nit-picky point that may only be relevant in a handful of cases. Overall I'd be happy with the current implementation.
> 
> * `4v5x`: you're right - bad choice by me this one indeed only has one assembly (but a ton of chains :D)
> * `6o3h`: Thank you for going in depth with this example. I agree that `"60"` would not be super concerning.
> * `1f2n`: This was the type of behaviour I was worried about, thank you for identifying one of these cases. For this particular one, the string only gets up to length `"X0-60"` , which isn't terrible but still pretty hefty compared to what it could be: stored e.g. via numpy's `<U` encoding -- utf32, i.e. 4 bytes iirc -- this would be `5*4=20` bytes per element. That's more than the `3*4=12` (float 32) bytes per coordinate in this case. Instead, with an integer encoding this could easily be `1` byte (uint8) or `2` bytes (uint16). This would make the atom-array both, more lightweight and comparisons over the `sym_id` dimension faster because we can do int comparisons rather than string comparisons.
> 
> What I don't know is if there are cases where this id could become even longer? (@padix-key or @nscorley, maybe you know more?). In the past I've often been surprised by obscure edge cases in the PDB that I thought would never happen, so nowadays I usually try to rely on as little assumptions on PDB files as possible.

@Croydon-Brixton In my view, either approach would work, but I tend to prefer the integer solution as well, for both the reasons that you and @padix-key suggested.

The longest symmetry ID I've seen (and thus a lower bound) is in `2fs3`: `'(X0)(1-10,21-25,31-35,41-50)'`. Given the number of atoms involved in this structure, when building the full biological assembly of up to a million atoms, such a field (a 23-character string) becomes non-negligible (back-of-the-envelope math gives me ~25mb, which is a bit on the larger side, especially given how lightweight the other annotation categories are).

Although converting to integers comes as the cost of losing the information about what transformations were applied to the corresponding chain in the asymmetric unit, I haven't come across any cases where that's needed for downstream uses.

[the remark this answers is not in the document]
Comment (padix-key) [reply src/biotite/structure/io/pdbx/convert.py:1669]: Actually the field would not be that long: `X0` would only be combined with one of the other operations, i.e. the sym IDs would be
```
X0-1, ..., X0-10, X0-21, ...
```
Nevertheless, as discussed above, using integers has still other advantages

**Review by padix-key — APPROVED** [review]

**Comment by padix-key** [comment]
I'll merge the PR now. Thanks again @alex-hh for implementing it and to @Croydon-Brixton and @nscorley for the feedback!

**Comment by Croydon-Brixton** [comment]
Always a pleasure! Thank you for all the great work on biotite!
